pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,180 @@
+	newServiceAccountTokenCommand.Flags().Int64Var(&options.TimeoutSeocnds, "timeout", 100, "the maximum time allowed to generate a token, in seconds")
+	return newServiceAccountTokenCommand
+}
+
+func (o *NewServiceAccountTokenOptions) Complete(args []string, f *clientcmd.Factory) error {
+	if len(args) != 1 {
+		return fmt.Errorf("expected one service account name as an argument, got %q", args)
+	}
+
+	o.SAName = args[0]
+
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+
+	if len(o.Namespace) == 0 {
+		namespace, _, err := f.DefaultNamespace()
+		if err != nil {
+			return fmt.Errorf("could not retrieve default namespace: %v", err)
+		}
+		o.Namespace = namespace
+	}
+
+	o.SAClient = client.ServiceAccounts(o.Namespace)
+	o.SecretsClient = client.Secrets(o.Namespace)
+	return nil
+}
+
+// Run creates a new token secret, waits for the service account token controller to fulfill it, then adds the token to the service account
+func (o *NewServiceAccountTokenOptions) Run() error {
+	serviceAccount, err := o.SAClient.Get(o.SAName)
+	if err != nil {
+		return fmt.Errorf("could not locate service account %q: %v", o.SAName, err)
+	}
+
+	tokenSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      secret.Strategy.GenerateName(osautil.GetTokenSecretNamePrefix(serviceAccount)),
+			Namespace: serviceAccount.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey: serviceAccount.Name,
+				api.ServiceAccountUIDKey:  string(serviceAccount.UID),
+			},
+		},
+		Type: api.SecretTypeServiceAccountToken,
+		Data: map[string][]byte{},
+	}
+
+	persistedToken, err := o.SecretsClient.Create(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	// we need to wait for the service account token controller to make the new token valid
+	options := api.ListOptions{
+		FieldSelector:   fields.SelectorFromSet(fields.Set(map[string]string{"type": string(api.SecretTypeServiceAccountToken)})),
+		Watch:           true,
+		ResourceVersion: persistedToken.ResourceVersion,
+		TimeoutSeconds:  &o.TimeoutSeocnds,
+	}
+	watcher, err := o.SecretsClient.Watch(options)
+	if err != nil {
+		return fmt.Errorf("could not begin watch for token: %v", err)
+	}
+	defer watcher.Stop()
+
+watch:
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok || event.Type == watch.Error {
+				return fmt.Errorf("encountered error while watching for token: %v", event.Object)
+			}
+
+			if eventToken, ok := event.Object.(*api.Secret); ok {
+				if eventToken.Name == persistedToken.Name && serviceaccounts.IsValidServiceAccountToken(serviceAccount, eventToken) && event.Type == watch.Modified {
+					fmt.Fprintf(o.Out, "token %q created for service account %q\n", eventToken.Name, serviceAccount.Name)
+					break watch
+				}
+			}
+		case <-time.After(time.Duration(o.TimeoutSeocnds) * time.Second):
```

Comment:
I thought this would only trigger on something coming out of the channel, i.e. the timer? What's the syntax for this?